src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -764,6 +796,116 @@ private final class BitmapIndexedMapNode[K, +V](
     else new BitmapIndexedMapNode[K, W](dataMap, nodeMap, newContent, originalHashes, size, cachedJavaKeySetHashCode)
   }
 
+  override def mergeInto[V1 >: V](that: MapNode[K, V1], builder: HashMapBuilder[K, V1], shift: Int)(mergef: ((K, V), (K, V1)) => (K, V1)): Unit = that match {
+    case bm: BitmapIndexedMapNode[K, V] =>
+      if (size == 0) {
+        that.buildTo(builder)
+        return
+      } else if (bm.size == 0) {
+        buildTo(builder)
+        return
+      }
+
+      val allMap = dataMap | bm.dataMap | nodeMap | bm.nodeMap
+
+      // minimumIndex is inclusive -- it is the first index for which there is data or nodes
+      val minimumBitPos: Int = Node.bitposFrom(Integer.numberOfTrailingZeros(allMap))
+      // maximumIndex is inclusive -- it is the last index for which there is data or nodes
+      // it could not be exclusive, because then upper bound in worst case (Node.BranchingFactor) would be out-of-bound
+      // of int bitposition representation
+      val maximumBitPos: Int = Node.bitposFrom(Node.BranchingFactor - Integer.numberOfLeadingZeros(allMap) - 1)
+
+      {
+        var bitpos = minimumBitPos
+        var leftIdx = 0
+        var rightIdx = 0
+        var finished = false
+
+        while (!finished) {
+
+          if ((bitpos & dataMap) != 0) {
+            val leftKey = getKey(leftIdx)
+            val leftValue = getValue(leftIdx)
+            val leftOriginalHash = getHash(leftIdx)
+            if ((bitpos & bm.dataMap) != 0) {
```

Comment:
Since the words `right data` or `right node` are used in several comments, it may be better to use those words in the code, rather than the comments,  by extracting some variables at the top-level of the `while`:  
```Scala
val rightData: Boolean = (bitpos & bm.dataMap) != 0 
val rightNode: Boolean = (bitpos & bm.nodeMap) != 0
```

Reply:
I guess the downside of that is that these cases are mutually exclusive -- if `rightData` then definitely not `rightNode`. But with this approach, you end up doing both computations anyways, whereas you don't in an `if (...) else if (...) else { ... }` approach. These computations likely are so fast that it doesn't matter though.